File: Leetcode/Binary Tree Inorder Traversal/iterative.cpp
// iterative solution to the problem
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<int> inorderTraversal(TreeNode *root)
    {
        stack<TreeNode *> traversal;
        vector<int> result;

        if (root == nullptr)
            return result;

        TreeNode *curr = root;

        while (curr != nullptr || !traversal.empty())
        {
            while (curr != nullptr)
            {
                traversal.push(curr);
                curr = curr->left;
            }

            curr = traversal.top();
            traversal.pop();
            result.push_back(curr->val);
            curr = curr->right;
        }

        return result;
    }
};

File: Leetcode/Binary Tree Inorder Traversal/Iterative.java
import java.util.*;

public class Iterative {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        LinkedList<Integer> result = new LinkedList<>();
        Stack<TreeNode> stack = new Stack<>();

        if (root == null)
            return result;

        TreeNode current = root;

        while (current != null || !stack.isEmpty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            current = stack.pop();
            result.add(current.val);

            current = current.right;
        }

        return result;
    }

}
